src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs

```diff
@@ -908,5 +915,296 @@ .maxstack  4
+public enum MyColor_UInt16 : ushort {{ Red, Orange, Yellow, Green, Blue }}
+public enum MyColor_Int16 : short {{ Red, Orange, Yellow, Green, Blue }}
+public enum MyColor_UInt32 : uint {{ Red, Orange, Yellow, Green, Blue }}
+public enum MyColor_Int32 : int {{ Red, Orange, Yellow, Green, Blue }}
+public enum MyColor_UInt64 : ulong {{ Red, Orange, Yellow, Green, Blue }}
+public enum MyColor_Int64 : long {{ Red, Orange, Yellow, Green, Blue }}
+
+public class Test
+{{
+    public static {type} M()
+    {{
+        System.ReadOnlySpan<{type}> s = new {type}[] {{ {args} }};
+        return s[0];
+    }}
+}}";
+
+            var compilation = CreateCompilationWithMscorlibAndSpan(csharp, TestOptions.ReleaseDll);
+            var verifier = CompileAndVerify(compilation, verify: Verification.Skipped);
+            verifier.VerifyIL("Test.M", @$"{{
+  // Code size       48 (0x30)
+  .maxstack  3
+  .locals init (System.ReadOnlySpan<{type}> V_0) //s
+  IL_0000:  ldsfld     ""{type}[] {fieldName}_Array""
+  IL_0005:  dup
+  IL_0006:  brtrue.s   IL_0020
+  IL_0008:  pop
+  IL_0009:  ldc.i4.{length}
+  IL_000a:  newarr     ""{type}""
+  IL_000f:  dup
+  IL_0010:  ldtoken    ""{fieldType} {fieldName}""
+  IL_0015:  call       ""void System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray(System.Array, System.RuntimeFieldHandle)""
+  IL_001a:  dup
+  IL_001b:  stsfld     ""{type}[] {fieldName}_Array""
+  IL_0020:  newobj     ""System.ReadOnlySpan<{type}>..ctor({type}[])""
+  IL_0025:  stloc.0
+  IL_0026:  ldloca.s   V_0
+  IL_0028:  ldc.i4.0
+  IL_0029:  call       ""ref readonly {type} System.ReadOnlySpan<{type}>.this[int].get""
+  IL_002e:  {ldind}
+  IL_002f:  ret
+}}");
+        }
+
+        [Theory]
+        [InlineData(0, 2)]
+        [InlineData(1, 1)]
+        [InlineData(1, 2)]
+        [InlineData(1, 3)]
+        [InlineData(0, 4)]
+        public void CreateSpanIsNotUsedForSpanOfNonSize1Types_NonFullLength(int start, int length)
+        {
+            var csharp = RuntimeHelpersCreateSpan + @$"
+public class Test
+{{
+    public static System.ReadOnlySpan<char> StaticData => new System.ReadOnlySpan<char>(new char[] {{ 'a', 'b', 'c' }}, {start}, {length});
+}}";
+            var compilation = CreateCompilationWithMscorlibAndSpan(csharp, TestOptions.ReleaseDll);
+            var verifier = CompileAndVerify(compilation, verify: Verification.Skipped);
+            verifier.VerifyIL("Test.StaticData.get", @$"{{
+  // Code size       29 (0x1d)
+  .maxstack  4
+  IL_0000:  ldc.i4.3
+  IL_0001:  newarr     ""char""
+  IL_0006:  dup
+  IL_0007:  ldc.i4.0
+  IL_0008:  ldc.i4.s   97
+  IL_000a:  stelem.i2
+  IL_000b:  dup
+  IL_000c:  ldc.i4.1
+  IL_000d:  ldc.i4.s   98
+  IL_000f:  stelem.i2
+  IL_0010:  dup
+  IL_0011:  ldc.i4.2
+  IL_0012:  ldc.i4.s   99
+  IL_0014:  stelem.i2
+  IL_0015:  ldc.i4.{start}
+  IL_0016:  ldc.i4.{length}
+  IL_0017:  newobj     ""System.ReadOnlySpan<char>..ctor(char[], int, int)""
```

Comment:
Let's also include a negative case, where the element size is not 1 and start/length arguments are provided (then the optimization doesn't kick in).